image/lytro_lfp.ksy

```diff
@@ -0,0 +1,158 @@
+            contents: [0x0d, 0x0a]
+          - id: unkn1
+            type: u2
+          - id: unkn0
+            type: u4
+          - id: size
+            type: u4
+          - id: identifier
+            size: 80
+            type: strz
+            doc: "Usually contains SHA-1 checksum"
+            if: size != 0
+          - id: data
+            size: size
+            type:
+              switch-on: type
+              cases:
+                #"'LFP'": header # usually empty?
+                "'LFM'": json_metadata
+                #"'LFC'": chunk # cannot process it yet
+        types:
+          json_metadata:
+            seq:
+              - id: json_str
+                size: _io.size
+                type: strz
+            instances:
+              parsed_metadata:
+                pos: 0
+                type: lytro_lfp_json_parsed_metadata(json_str, _root)
+                doc: |
+                  # put to lytro_lfp_json_parsed_metadata.py
+                  import json
+                  from kaitaistruct import KaitaiStruct
+                  from enum import IntEnum
+                  
+                  class ChunkType(IntEnum):
+                    metadata = 0
+                    private_metadata = 1
+                    image = 2
+                  
+                  class Frame:
+                    __slots__ = ("image", "metadata", "private_metadata")
+                    def __init__(self, jsonObj, index):
+                      self.metadata = index[jsonObj['metadataRef']]
+                      
+                      self.private_metadata = index[jsonObj['privateMetadataRef']]
+                      self.image = index[jsonObj['imageRef']]
+                      
+                  
+                  class SectionLink:
+                    __slots__ = ("type", "frame_no")
+                    def __init__(self, type, frame_no):
+                      self.type = type
+                      self.frame_no = frame_no
+                  
+                  class LytroLfpJsonParsedMetadata(KaitaiStruct):
+                    __slots__ = ("json", "frames", "version", "metadata_index")
+                    def __init__(self, json_str: str, root, _io, _parent=None, _root=None):
+                      self.json = json.loads(json_str)
+                      index = root.index
+                      pic = self.json['picture']
+                      self.frames = []
+                      self.metadata_index = {}
+                      
+                      for i, f in enumerate(pic['frameArray']):
+                        fDescr = Frame(f['frame'], index)
+                        self.frames.append(fDescr)
+                        self.metadata_index[fDescr.metadata] = SectionLink(i, ChunkType.metadata)
+                        self.metadata_index[fDescr.private_metadata] = SectionLink(i, ChunkType.private_metadata)
+                        self.metadata_index[fDescr.image] = SectionLink(i, ChunkType.image)
```

Comment:
It's kind of pointless to merge this into main KS repo, as this clearly requires these external modules, and (1) all the current infrastructure does not really support that, (2) I doubt that it's a good way to go for future development — the whole point of having KS specs was having cross-language support, and this particular implementation will be constrained to a single language.

Reply:
>and this particular implementation will be constrained to a single language.

It is an example of missing pieces in python. It is pointless to insert there impls in all other langs. A user should look at python code and implement something similar in the lang he uses. The rest of the spec is suitable for other langs (not necessarily all of them).

Reply:
I don't know if Python is a good language for that. There's a library I found where only a Python implementation existed, and it is becoming the _hardest thing ever_ to port to Java.

Reply:
As we already know, the proper solution requres changes in KSC. But for now a temporary solution working at least somewhere is better than nothing.